src/mpyl/stages/discovery.py

```diff
@@ -227,39 +221,31 @@ def build_project_execution(
             logger.debug(
                 f"Project {project} will execute stage {stage} because a dependency was modified"
             )
-            # pylint: disable=fixme
-            # FIXME ptab we should still calculate a hash (in case files changed) to speed up a possible next build
-            # this will come in a following PR
-            return ProjectExecution.always_run(project)
+
+            if is_project_modified:
+                hashed_changes = _hash_changes_in_project(
+                    logger=logger, project=project, changeset=changeset
+                )
+            else:
+                hashed_changes = None
+
+            return ProjectExecution.run(project, hashed_changes)
 
         if is_project_modified:
-            cache_key = _cache_key_from_changes_in_project(
+            hashed_changes = _hash_changes_in_project(
                 logger=logger, project=project, changeset=changeset
             )
-            if cache_key:
-                logger.debug(
-                    f"Project {project.name}: using hash of modified files as cache key {cache_key}"
-                )
-            else:
-                # pylint: disable=fixme
-                # FIXME ptab using revision as a cache_key makes no sense, as we will never check against
-                #  that commit against. Would be simpler just to fallback to the UUID generation here for consistency.
-                logger.debug(
-                    f"Project {project.name}: no content changes, falling back to revision as"
-                    f" cache key: {changeset.sha}"
-                )
-                cache_key = changeset.sha
 
-            return ProjectExecution(
+            return ProjectExecution.create(
                 project=project,
-                cache_key=cache_key,
                 cached=is_project_cached_for_stage(
```

Comment:
Ideally we would move this method inside `ProjectExecution`, however that creates a circular reference with `project_execution.py` → `models.py` → `project_execution.py` so we can't easily do that. I might deal with that in a follow-up PR.